crypto/uid.c

```diff
@@ -31,8 +31,20 @@ int OPENSSL_issetugid(void)
 # include OPENSSL_UNISTD
 # include <sys/types.h>
 
+# if defined(__GLIBC__)
+#  if defined(__GLIBC_PREREQ)
+#   if __GLIBC_PREREQ(2, 16)
+#    include <sys/auxv.h>
+#   endif
+#  endif
+# endif
+
 int OPENSSL_issetugid(void)
 {
+# ifdef AT_SECURE
+    if (getauxval(AT_SECURE) != 0)
+        return 1;
+# endif
```

Comment:
In the `getauxval(AT_SECURE) == 0` case, is it really necessary to do the `getuid()` and `getgid()` tests as fallback? Don't we trust `getauxval(AT_SECURE)`? Couldn't we just replace the latter tests with the former if if it exists?

```
# ifdef AT_SECURE
    if (getauxval(AT_SECURE) != 0)
        return 1;
# else
    if (getuid() != geteuid())
        return 1;
    if (getgid() != getegid())
         return 1;
# endif
     return 0;
```

Reply:
You could also `return getauxval(AT_SECURE) != 0;`

But I really don't have a preference either, just pick one.

Reply:
I like Kurt's suggestion, but if you choose it, you should do it for the else part, too:

```
# ifdef AT_SECURE
    return getauxval(AT_SECURE) != 0;
# else
    return getuid() != geteuid() || getgid() != getegid();
# endif
```

Reply:
I'm going with the final suggestion by @mspncp here.